packages/devops/scripts/lambda/utilities.py

```diff
@@ -0,0 +1,236 @@
+elbv2 = boto3.client('elbv2')
+acm = boto3.client('acm')
+
+
+def get_tag(instance, tag_name):
+    try:
+        tag_value = [
+            t.get('Value') for t in instance['Tags']
+            if t['Key'] == tag_name][0]
+    except IndexError:
+        tag_value = ''
+    return tag_value
+
+
+def get_instances(filters):
+    reservations = ec.describe_instances(
+        Filters=filters
+    ).get(
+        'Reservations', []
+    )
+    return reservations[0]['Instances']
+
+
+def random_string(length):
+    return ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(length))
+
+
+def tags_contains(tags, key, value):
+    tags_matching_key = list(filter(lambda x: x['Key'] == key, tags))
+    if len(tags_matching_key) == 0:
+        return False
+    tag_matching_key = tags_matching_key[0]
+    return tag_matching_key['Value'] == value
+
+
+def get_cert(type_tag):
+    certs = acm.list_certificates(MaxItems=400)
+    for cert in certs['CertificateSummaryList']:
+        tags = acm.list_tags_for_certificate(CertificateArn=cert['CertificateArn'])
+        if tags_contains(tags['Tags'], 'Type', type_tag):
+            return cert
+    raise Exception('No certificate found with type tag: ' + type_tag)
+
+
+def get_gateway_elb(tupaia_instance_name):
+    elbs = elbv2.describe_load_balancers(PageSize=400)
+    tags_by_elb = elbv2.describe_tags(
+        ResourceArns=list(map(lambda x: x['LoadBalancerArn'], elbs['LoadBalancers']))
+    )
+    matching_arns = list(filter(lambda x: tags_contains(x['Tags'], 'Type', 'Tupaia-Gateway') and tags_contains(x['Tags'], 'TupaiaInstanceName', tupaia_instance_name), tags_by_elb['TagDescriptions']))
+    if len(matching_arns) == 0:
+        raise Exception('No gateway elb found tagged with tupaia instance name: ' + tupaia_instance_name)
+    matching_arn = matching_arns[0]['ResourceArn']
+    return list(filter(lambda x: x['LoadBalancerArn'] == matching_arn, elbs['LoadBalancers']))[0]
+
+
+def get_gateway_target_group(tupaia_instance_name):
+    target_groups = elbv2.describe_target_groups(PageSize=400)
+    tags_by_target_group = elbv2.describe_tags(
+        ResourceArns=list(map(lambda x: x['TargetGroupArn'], target_groups['TargetGroups']))
+    )
+    matching_arns = list(filter(lambda x: tags_contains(x['Tags'], 'Type', 'Tupaia-Gateway') and tags_contains(x['Tags'], 'TupaiaInstanceName', tupaia_instance_name), tags_by_target_group['TagDescriptions']))
+    if len(matching_arns) == 0:
+        raise Exception('No gateway target group found tagged with tupaia instance name: ' + tupaia_instance_name)
+    matching_arn = matching_arns[0]['ResourceArn']
+    return list(filter(lambda x: x['TargetGroupArn'] == matching_arn, target_groups['TargetGroups']))[0]
+
+
+def get_gateway_listeners(tupaia_instance_name):
+    elb = get_gateway_elb(tupaia_instance_name)
+    return elbv2.describe_listeners(
+        LoadBalancerArn=elb['LoadBalancerArn'],
+        PageSize=100
+    )['Listeners']
```

Comment:
I think you might be able to follow this same pattern for `get_gateway_target_group`. Seems a bit simpler and removes a dependency on having tags set up correctly at that lower level (though I think that's quite helpful for manual diagnostic purposes)

Reply:
I think I considered this and wanted to keep the lookup by tags. 

The issue is that because target groups and ELBs are separate resources, they might not be a 1:1 mapping in the future. 

So if you did the target group lookup by ELB you might get multiple target groups for the same ELB, one for the gateway and one for something else.